File: benchmarks/cross_lang/typescript/rb_tree.ts
// Functional-style red-black tree port for cross-language benchmark.
// Mirrors examples/perceus/rb_tree.kai. Objects play the role of immutable
// nodes; the runtime GC handles memory just like Perceus does at compile time.

const RED = 0;
const BLACK = 1;

type Node = {
  color: number;
  left: Node | null;
  key: number;
  value: number;
  right: Node | null;
};

function mk(c: number, l: Node | null, k: number, v: number, r: Node | null): Node {
  return { color: c, left: l, key: k, value: v, right: r };
}

function balance(c: number, l: Node | null, k: number, v: number, r: Node | null): Node {
  if (c === BLACK) {
    if (l !== null && l.color === RED) {
      const ll = l.left;
      if (ll !== null && ll.color === RED) {
        return mk(RED,
                  mk(BLACK, ll.left, ll.key, ll.value, ll.right),
                  l.key, l.value,
                  mk(BLACK, l.right, k, v, r));
      }
      const lr = l.right;
      if (lr !== null && lr.color === RED) {
        return mk(RED,
                  mk(BLACK, l.left, l.key, l.value, lr.left),
                  lr.key, lr.value,
                  mk(BLACK, lr.right, k, v, r));
      }
    }
    if (r !== null && r.color === RED) {
      const rl = r.left;
      if (rl !== null && rl.color === RED) {
        return mk(RED,
                  mk(BLACK, l, k, v, rl.left),
                  rl.key, rl.value,
                  mk(BLACK, rl.right, r.key, r.value, r.right));
      }
      const rr = r.right;
      if (rr !== null && rr.color === RED) {
        return mk(RED,
                  mk(BLACK, l, k, v, r.left),
                  r.key, r.value,
                  mk(BLACK, rr.left, rr.key, rr.value, rr.right));
      }
    }
  }
  return mk(c, l, k, v, r);
}

function insertLoop(t: Node | null, k: number, v: number): Node {
  if (t === null) return mk(RED, null, k, v, null);
  if (k < t.key) return balance(t.color, insertLoop(t.left, k, v), t.key, t.value, t.right);
  if (k > t.key) return balance(t.color, t.left, t.key, t.value, insertLoop(t.right, k, v));
  return mk(t.color, t.left, t.key, v, t.right);
}

function rbInsert(t: Node | null, k: number, v: number): Node {
  const r = insertLoop(t, k, v);
  return mk(BLACK, r.left, r.key, r.value, r.right);
}

function rbSize(t: Node | null): number {
  if (t === null) return 0;
  return 1 + rbSize(t.left) + rbSize(t.right);
}

function rbHeight(t: Node | null): number {
  if (t === null) return 0;
  const l = rbHeight(t.left);
  const r = rbHeight(t.right);
  return 1 + (l > r ? l : r);
}

function lcgNext(s: number): number {
  // (s * 1664525 + 1013904223) mod (2^31 - 1). Safe in number (<= 2^53).
  return (s * 1664525 + 1013904223) % 2147483647;
}

function now(): number {
  // Hi-res clock — process.hrtime.bigint on Node, performance.now on Deno.
  // Use performance.now (ms) where available; fall back to Date.now.
  if (typeof performance !== "undefined" && performance.now) {
    return performance.now();
  }
  return Date.now();
}

function main() {
  let root: Node | null = null;
  let seed = 1;
  const n = 1000000;

  const startMs = now();
  let i = n;
  while (i > 0) {
    seed = lcgNext(seed);
    root = rbInsert(root, seed, i);
    i -= 1;
  }
  const elapsedMs = now() - startMs;

  const secs = Math.floor(elapsedMs / 1000);
  const ms = Math.floor(elapsedMs) % 1000;

  const label = typeof Deno !== "undefined"
    ? "deno (functional RB)"
    : "node (functional RB)";
  console.log(`${label} rb-tree benchmark`);
  console.log(`inserts: ${n}`);
  console.log(`size: ${rbSize(root)}`);
  console.log(`height: ${rbHeight(root)}`);
  console.log(`elapsed: ${secs}.${String(ms).padStart(3, "0")}s`);
}

// @ts-ignore — Deno global is optional.
declare const Deno: unknown;
main();

File: benchmarks/cross_lang/go/rb_tree.go
// Functional-style red-black tree port for cross-language benchmark.
// Mirrors examples/perceus/rb_tree.kai. Uses pointer-shared sub-trees
// (Go's GC stands in for kaikai's Perceus reuse-in-place at the
// memory-discipline level — both produce a directed graph of shared
// nodes with no manual frees).
package main

import (
	"fmt"
	"time"
)

type Color int

const (
	Red Color = iota
	Black
)

type Tree struct {
	color Color
	left  *Tree
	key   int64
	value int64
	right *Tree
}

func mk(c Color, l *Tree, k int64, v int64, r *Tree) *Tree {
	return &Tree{color: c, left: l, key: k, value: v, right: r}
}

func balance(c Color, l *Tree, k int64, v int64, r *Tree) *Tree {
	if c == Black {
		if l != nil && l.color == Red {
			if l.left != nil && l.left.color == Red {
				ll := l.left
				return mk(Red,
					mk(Black, ll.left, ll.key, ll.value, ll.right),
					l.key, l.value,
					mk(Black, l.right, k, v, r))
			}
			if l.right != nil && l.right.color == Red {
				lr := l.right
				return mk(Red,
					mk(Black, l.left, l.key, l.value, lr.left),
					lr.key, lr.value,
					mk(Black, lr.right, k, v, r))
			}
		}
		if r != nil && r.color == Red {
			if r.left != nil && r.left.color == Red {
				rl := r.left
				return mk(Red,
					mk(Black, l, k, v, rl.left),
					rl.key, rl.value,
					mk(Black, rl.right, r.key, r.value, r.right))
			}
			if r.right != nil && r.right.color == Red {
				rr := r.right
				return mk(Red,
					mk(Black, l, k, v, r.left),
					r.key, r.value,
					mk(Black, rr.left, rr.key, rr.value, rr.right))
			}
		}
	}
	return mk(c, l, k, v, r)
}

func insertLoop(t *Tree, k int64, v int64) *Tree {
	if t == nil {
		return mk(Red, nil, k, v, nil)
	}
	if k < t.key {
		return balance(t.color, insertLoop(t.left, k, v), t.key, t.value, t.right)
	}
	if k > t.key {
		return balance(t.color, t.left, t.key, t.value, insertLoop(t.right, k, v))
	}
	return mk(t.color, t.left, t.key, v, t.right)
}

func rbInsert(t *Tree, k int64, v int64) *Tree {
	r := insertLoop(t, k, v)
	if r == nil {
		return nil
	}
	return mk(Black, r.left, r.key, r.value, r.right)
}

func rbSize(t *Tree) int64 {
	if t == nil {
		return 0
	}
	return 1 + rbSize(t.left) + rbSize(t.right)
}

func rbHeight(t *Tree) int64 {
	if t == nil {
		return 0
	}
	l := rbHeight(t.left)
	r := rbHeight(t.right)
	if l > r {
		return 1 + l
	}
	return 1 + r
}

func lcgNext(s int64) int64 {
	return (s*1664525 + 1013904223) % 2147483647
}

func main() {
	var root *Tree
	var seed int64 = 1
	var n int64 = 1000000

	start := time.Now()
	for i := n; i > 0; i-- {
		seed = lcgNext(seed)
		root = rbInsert(root, seed, i)
	}
	elapsed := time.Since(start)

	secs := elapsed.Milliseconds() / 1000
	ms := elapsed.Milliseconds() % 1000

	fmt.Println("go (functional RB) rb-tree benchmark")
	fmt.Printf("inserts: %d\n", n)
	fmt.Printf("size: %d\n", rbSize(root))
	fmt.Printf("height: %d\n", rbHeight(root))
	fmt.Printf("elapsed: %d.%03ds\n", secs, ms)
}
